docs/proposals/image-promotion.md

```diff
@@ -0,0 +1,152 @@
+* Remote devops team pull the image from shared Docker registry and make it available in "stage" cluster.
+* Remote devops team promotes the image from "stage" to "production"
+
+### Manual promotion of builder image using external Docker registry
+
+As a platform operator, I want to be able to promote the S2I builder image(s) when a new
+version is available in Red Hat Docker registry. I want to perform additional verifications
+(iow. deploy to "stage" first) before rolling a new image out to production which causes rebuild
+of thousand applications.
+
+#### Example
+
+* OpenShift engineering push the `openshift/ruby-22-rhel7` image to internal CI registry.
+* OpenShift QA team pull the image from the internal CI registry and run tests.
+* OpenShift QA team sign-off the Docker image when tests pass and push it to "integration" registry.
+* OpenShift operations team pulls the image from "integration" registry and promote it to "stage" environment.
+* OpenShift operations team manually promote the image to production.
+
+### Semi-automatic promotion and testing of builder images
+
+As a OpenShift S2I image developer, I want to have the image rebuilt, tested and pushed to
+the Docker Hub or to internal registry and tag it as "qa-ready" and notify the QA team about
+availability of a new Docker Image.
+
+#### Example of pull request tester
+
+* OpenShift engineering propose a change to `openshift/ruby-22-rhel7` image.
+* CI starts OpenShift server and execute the extended tests cases related to this image.
+* CI send notification to original pull request.
+
+#### Example of push_images job
+
+* OpenShift engineering merge change to `openshift/ruby-22-rhel7` repository.
+* OpenShift server detect the change and execute test suite for given image.
+* If the image is 'base' image, OpenShift server launch multiple tests for all dependent images.
+* OpenShift server push the image to internal registry and Docker Hub (see: "Manual promotion of builder image")
+
+## Areas of improvement
+
+The first use-case describes a typical Continuos Delivery example, where the developer uses
+both the Jenkins server and the OpenShift to deliver updates to the application.
+
+### Jenkins
+
+* There is no support for enabling the Jenkins when creating a new application.
+* There is no official "jenkins" plugin for OpenShift available.
+* Users have to manually tweak the Jenkins job parameters.
+* The Jenkins job definition is written in shell and it is complex.
+
+### Deployments
+
+* The "post" lifecycle hook can be executed only in the image that is being deployed.
+* The images must have OpenShift CLI tools installed in order to do tagging.
+* The container must have OpenShift API secrets mounted.
+* There is no way to say "scale this deployment down" after verification finishes
+  (must be done in verification script).
+
+### Builds
+
+* Allow to push image to multiple Docker registries
```

Comment:
Agreed

On Aug 5, 2015, at 11:41 PM, Ben Parees <notifications@github.com> wrote:

In docs/proposals/image-promotion.md
<https://github.com/openshift/origin/pull/4016#discussion_r36378971>:

> +* There is no support for enabling the Jenkins when creating a new application.
> +* There is no official "jenkins" plugin for OpenShift available.
> +* Users have to manually tweak the Jenkins job parameters.
> +* The Jenkins job definition is written in shell and it is complex.
> +
> +### Deployments
> +
> +* The "post" lifecycle hook can be executed only in the image that is being deployed.
> +* The images must have OpenShift CLI tools installed in order to do tagging.
> +* The container must have OpenShift API secrets mounted.
> +* There is no way to say "scale this deployment down" after verification finishes
> +  (must be done in verification script).
> +
> +### Builds
> +
> +* Allow to push image to multiple Docker registries

not sure i see the actual use case here, from a CI/CD perspective? even our
in-house use case really amounts to the fact that we're doing two builds as
if they are one (rhel and centos) so in openshift we'd probably just set it
up as 2 buildconfigs. We do push to 2 places but you can argue that that's
really promotion......ie we should build the rhel and centos images with
them outputting to a single internal repo. Then we should test those images
and, if they pass, promote them to multiple destination registries if
desired.

but i don't think the build itself should push it to multiple locations.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4016/files#r36378971>.